src/charm.py

```diff
@@ -3,113 +3,97 @@
-            args: Arguments passed to the CharmBase parent constructor.
+            args: Arguments to initialize the charm base.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        try:
+            self.state = State.from_charm(self)
+        except InvalidStateError as e:
+            logger.debug("Error parsing charm_state %s", e)
+            self.unit.status = ops.BlockedStatus(e.msg)
+            return
+
+        self.jenkins_agent_service = service.JenkinsAgentService(self.state)
+        self.agent_observer = agent_observer.Observer(self, self.state, self.jenkins_agent_service)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.start, self.reconcile)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent):
-        """Define and start a workload using the Pebble API.
-
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
-
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
-
-        Args:
-            event: event triggering the handler.
-        """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
-
-    def _on_config_changed(self, event: ops.ConfigChangedEvent):
-        """Handle changed configuration.
-
-        Change this example to suit your needs. If you don't need to handle config, you can remove
-        this method.
-
-        Learn more about config at https://juju.is/docs/sdk/config
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+    def _on_install(self, _: ops.InstallEvent) -> None:
+        """Handle install event, setup the agent service."""
+        try:
+            self.jenkins_agent_service.install()
+        except service.PackageInstallError as e:
+            logger.debug("Error installing the agent service %s", e)
+            self.unit.status = ops.ErrorStatus("Error installing the agent service")
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle config changed event. Update the agent's label in the relation's databag."""
+        if agent_relation := self.model.get_relation(AGENT_RELATION):
```

Comment:
I think this syntax might not work on 3.8 🙁 - we need to make sure the tests run on both 3.8 and 3.10, i.e., all the Python version associated with the Ubuntu OS we support

Reply:
Indeed, I mistook it for something else